### PR TITLE
Ensure PackageInstallerService does not require the UI thread during GetSuggestedActions

### PR DESCRIPTION
**Pull request by sharwell**
Supersedes #43630

### CODE REVIEW

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 1]
```diff
@@ -103,21 +104,27 @@ public ImmutableArray<PackageSource> GetPackageSources()
...
+            else if (_packageSourcesAsync.IsCompleted)
             {
-                // This exception can happen when the nuget.config file is broken, e.g. invalid credentials.
-                // https://github.com/dotnet/roslyn/issues/40857
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                packageSources = _packageSourcesAsync.GetAwaiter().GetResult();
+            }
+            else
+            {
+                // Return without caching a result
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:126]: doc the impact.  primarily that some features may show incomplete resulsts for a short period of time, but that that's acceptable.

File: src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs [hunk 2]
```diff
@@ -57,7 +57,7 @@ public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
 
             var installerService = GetPackageInstallerService(document);
             var packageSources = searchNuGetPackages && symbolSearchService != null && installerService?.IsEnabled(document.Project.Id) == true
-                ? installerService.GetPackageSources()
+                ? await installerService.GetPackageSourcesAsync(allowSwitchToMainThread: false, context.CancellationToken).ConfigureAwait(false)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs:60]: all callers seem to pass 'false', so not sure if we actually need the parameter.  i think it's fine to say: this may not return anything in some cases.

File: src/Workspaces/Core/Portable/Packaging/IPackageInstallerService.cs [hunk 3]
```diff
@@ -28,7 +29,7 @@ internal interface IPackageInstallerService : IWorkspaceService
         bool CanShowManagePackagesDialog();
         void ShowManagePackagesDialog(string packageName);
 
-        ImmutableArray<PackageSource> GetPackageSources();
+        ValueTask<ImmutableArray<PackageSource>> GetPackageSourcesAsync(bool allowSwitchToMainThread, CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Packaging/IPackageInstallerService.cs:32]: maybe TryGetPackageSourcesAsync and doc this may be initialized in the bG and may not return results initially but may return results later.

**Comment by CyrusNajmabadi** [comment]
LGTM.  i like this approach.  esp. that it doesn't allow JTF in extensions.

File: src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs (hunk 2 above)
Comment (sharwell) [reply src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs:60]: I prefer to have the parameter as a highlight to potentially-inconvenient behavior. If the service becomes free-threaded in the future we can eliminate the parameter.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: Done

File: src/Workspaces/Core/Portable/Packaging/IPackageInstallerService.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Packaging/IPackageInstallerService.cs]: Updated

File: src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs]: so the reason i don't love that is simply because we have a codepath that is effectively untested.  so i don't really know if it actually works or not :)

File: src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs [hunk 4]
```diff
@@ -57,11 +57,19 @@ public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
 
             var installerService = GetPackageInstallerService(document);
             var packageSources = searchNuGetPackages && symbolSearchService != null && installerService?.IsEnabled(document.Project.Id) == true
-                ? installerService.GetPackageSources()
+                ? await installerService.TryGetPackageSourcesAsync(allowSwitchToMainThread: false, context.CancellationToken).ConfigureAwait(false)
                 : ImmutableArray<PackageSource>.Empty;
 
+            if (packageSources is null)
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs:63]: #nullable enable the file so we have the tracking for nullable value types too?

File: src/Features/Core/Portable/AddPackage/AbstractAddPackageCodeFixProvider.cs [hunk 5]
```diff
@@ -58,7 +58,7 @@ internal abstract partial class AbstractAddPackageCodeFixProvider : CodeFixProvi
                 searchNugetPackages &&
                 installerService.IsEnabled(document.Project.Id))
             {
-                foreach (var packageSource in installerService.GetPackageSources())
+                foreach (var packageSource in await installerService.TryGetPackageSourcesAsync(allowSwitchToMainThread: false, cancellationToken).ConfigureAwait(false))
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/AddPackage/AbstractAddPackageCodeFixProvider.cs:61]: Isn't this missing the null check? It seems the compiler lets you enumerate a Nullable<ImmutableArray<...>> and implicitly does the .Value for you, but that'll throw at runtime if it's null.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 6]
```diff
@@ -103,21 +104,28 @@ public ImmutableArray<PackageSource> GetPackageSources()
                 return packageSources;
             }
 
-            try
+            lock (_gate)
             {
-                packageSources = _packageSourceProvider.Value.GetSources(includeUnOfficial: true, includeDisabled: false)
-                    .SelectAsArray(r => new PackageSource(r.Key, r.Value));
+                if (_packageSourcesAsync is null)
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:109]: `#nullable enable` the file?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 7]
```diff
@@ -103,21 +104,28 @@ public ImmutableArray<PackageSource> GetPackageSources()
...
+                }
             }
-            catch (Exception ex) when (ex is InvalidDataException || ex is InvalidOperationException)
+
+            if (allowSwitchToMainThread)
             {
-                // These exceptions can happen when the nuget.config file is broken.
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                using var combinedToken = _tokenSource.Token.CombineWith(cancellationToken);
+                packageSources = await _packageSourcesAsync.JoinAsync(combinedToken.Token).ConfigureAwait(false);
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:118]: If we're allowing a switch to the main thread but a workspace change happens because the UI thread isn't blocked, can't this result in the request being cancelled because _tokenSource.Token will be cancelled, even though the caller won't expect it?

Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:118]: A agree we shouldn't be cancelling the operation except when `cancellationToken` is canceled. But what do we do for this case?

File: src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs (hunk 4 above)
Comment (sharwell) [reply src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs:63]: I'm leaving this one for a future change

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 6 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: Leaving this one for a future change. The one case where it looked like it could have helped us isn't caught by the compiler anyway.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 7 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: This is now fixed.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 8]
```diff
@@ -103,21 +104,46 @@ public ImmutableArray<PackageSource> GetPackageSources()
...
+            if (allowSwitchToMainThread)
             {
-                // These exceptions can happen when the nuget.config file is broken.
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                try
+                {
+                    using var combinedToken = _tokenSource.Token.CombineWith(cancellationToken);
+                    packageSources = await _packageSourcesAsync.JoinAsync(combinedToken.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:125]: This `catch` block had become unreachable in a previous change. I'm removing it.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 9]
```diff
@@ -130,6 +156,55 @@ public ImmutableArray<PackageSource> GetPackageSources()
...
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    // The failure may have been caused by a workspace change; try again after a short delay
+                    await Task.Delay(TimeSpan.FromMilliseconds(250)).ConfigureAwait(false);
+                }
+            }
+
+            // Local function
+            async Task<ImmutableArray<PackageSource>> GetPackageSourcesImplAsync(CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:185]: It's not possible. All the tokens have the same variable name.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 10]
```diff
@@ -130,6 +156,55 @@ public ImmutableArray<PackageSource> GetPackageSources()
...
+            // Automatically retry the operation if it gets cancelled due to a change in the workspace.
+            while (true)
+            {
+                var cancellationToken = _tokenSource.Token;
+                try
+                {
+                    var result = await GetPackageSourcesImplAsync(cancellationToken).ConfigureAwait(false);
+                    if (result.IsEmpty && cancellationToken.IsCancellationRequested)
+                    {
+                        // The failure may have been caused by a workspace change; try again after a short delay
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:170]: Sure, it could fail with `InvalidDataException` _and then_ have a workspace change cancel the token.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 11]
```diff
@@ -130,6 +156,55 @@ public ImmutableArray<PackageSource> GetPackageSources()
...
+        private async Task<ImmutableArray<PackageSource>> GetPackageSourcesImplAsync()
+        {
+            // Automatically retry the operation if it gets cancelled due to a change in the workspace.
+            while (true)
+            {
+                var cancellationToken = _tokenSource.Token;
+                try
+                {
+                    var result = await GetPackageSourcesImplAsync(cancellationToken).ConfigureAwait(false);
+                    if (result.IsEmpty && cancellationToken.IsCancellationRequested)
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:168]: > Why are we handling both this and also the OperationCancelledException in the catch block?

Cancelling a token may result in an exception, but this is never a guarantee (and rarely a guarantee even in cases where the token is cancelled before the start of the call). Cancellation tokens are frequently checked after calls.

> if the only thing actually using the cancellationToken is the SwitchToMainThreadAsync() then I don't see how this could happen in a way we'd care about?

I don't understand this part of the comment. Can you rephrase?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 12]
```diff
@@ -130,6 +156,55 @@ public ImmutableArray<PackageSource> GetPackageSources()
...
+                    {
+                        // The failure may have been caused by a workspace change; try again after a short delay
+                        await Task.Delay(TimeSpan.FromMilliseconds(250)).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:171]: The package sources are only retrieved once. They are not reloaded when the workspace changes.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 13]
```diff
@@ -103,21 +104,37 @@ public ImmutableArray<PackageSource> GetPackageSources()
...
+                }
             }
-            catch (Exception ex) when (ex is InvalidDataException || ex is InvalidOperationException)
+
+            if (allowSwitchToMainThread)
             {
-                // These exceptions can happen when the nuget.config file is broken.
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                using var combinedToken = _tokenSource.Token.CombineWith(cancellationToken);
+                packageSources = await _packageSourcesAsync.JoinAsync(combinedToken.Token).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:121]: why does this end up switching to the main thread?

Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:121]: STA COM

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 14]
```diff
@@ -103,21 +104,37 @@ public ImmutableArray<PackageSource> GetPackageSources()
...
-                // This exception can happen when the nuget.config file is broken, e.g. invalid credentials.
-                // https://github.com/dotnet/roslyn/issues/40857
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                if (_packageSourcesAsync.Task.Status != TaskStatus.RanToCompletion)
+                {
+                    // The operation did not complete successfully
+                    return null;
+                }
+
+                packageSources = _packageSourcesAsync.GetAwaiter().GetResult();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:131]: do we need the above 'if' check?  normally we would 'await' this, meaning we would get exceptions. so .GetAwaiter().GetResult() in all cases seems reasonable.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:162]: really?  that seems super strange to me.  i thought we listneed for the events and cleared the value, causing it to be required.

[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:162]: 📝 This is not a behavior change from the previous implementation.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 14 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:131]: Makes sense. The check was added prior to the automatic retry in `GetPackageSourcesImplAsync`.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:162]: So we have this code today:

```c#
        private void OnSourceProviderSourcesChanged(object sender, EventArgs e)
        {
            _packageSources = default;
            PackageSourcesChanged?.Invoke(this, EventArgs.Empty);
        }
```

which is intended for clearing the values we've cached when teh user changes them in VS>

[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:162]: Will check. I was thinking specifically of the case where the sources change in NuGet.config.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 14 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: Simplified to use the `await` for this scenario.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:143]: Fixed (restored ability to update this service when the user changes package sources) by eliminating `_packageSources` and using `_packageSourcesAsync` as the original source of truth.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:158]: Today they do. Tomorrow is what worries me. :smile:

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 15]
```diff
@@ -93,41 +93,89 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
...
         }
 
-        public ImmutableArray<PackageSource> GetPackageSources()
+        public async ValueTask<ImmutableArray<PackageSource>?> TryGetPackageSourcesAsync(bool allowSwitchToMainThread, CancellationToken cancellationToken)
         {
-            // Only read from _packageSources once, since OnSourceProviderSourcesChanged could reset it to default at
-            // any time while this method is running.
-            var packageSources = _packageSources;
-            if (packageSources != null)
+            // Only read from _packageSourcesAsync once, since OnSourceProviderSourcesChanged could reset it to default
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:98]: This _looks_ like a violation of the comment given it's doing a double-checked lock pattern here. Do we need the extra complexity versus just deleting the outer check? This isn't a high-traffic lock.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:141]: > Cancellation tokens are frequently checked after calls.

I guess I'd prefer it was just cancellationToken.ThrowIfCancellationRequested() which can jump to the catch block; that way you're not duplicating the retry logic.

[the remark this answers is not in the document]
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:141]: > if the only thing actually using the cancellationToken is the SwitchToMainThreadAsync() then I don't see how this could happen in a way we'd care about?

As you observed on my other question the only way we can hit this is if it failed, and then we had the badly timed workspace change. But why does that combination require special handling like this? I would have assumed either:

1. If it fails, it just failed, and whether we failed to observe the cancellation is moot -- it's broken either way. Either we accept the brokenness or want to retry, period.
2. We had cancelled, but if we got back a successful result, why not just use it? We didn't want it, but if you've already done the work why not keep it?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 16]
```diff
@@ -93,41 +93,89 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
...
+                // Since the task is already completed, we know this 'await' will complete synchronously.
+                return await packageSourcesAsync;
             }
-            catch (Exception ex) when (ex is InvalidDataException || ex is InvalidOperationException)
+            else if (allowSwitchToMainThread)
             {
-                // These exceptions can happen when the nuget.config file is broken.
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                using var combinedToken = _tokenSource.Token.CombineWith(cancellationToken);
+                return await packageSourcesAsync.JoinAsync(combinedToken.Token).ConfigureAwait(false);
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:122]: Isn't the original bug I blocked for back here?

https://github.com/dotnet/roslyn/pull/43718#discussion_r418415466

Because this now means a caller passing allowSwitchToMainThread = true can get cancelled due to the unexpected workspace change?

(I'm very much with @CyrusNajmabadi in that if this code path isn't active, let's just delete it.)

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:122]: I'm not sure if the intent was packageSourcesAsync now _never_ cancels due to workspace changes, but will now just retry until it finally suceeds, but then the _tokenSource.Token.CombineWith can probably go away?

Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:122]: The `CombineWith` is no longer necessary here. Will remove.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:158]: This is not a problem unique to this pull request. I would recommend filing an issue in dotnet/roslyn-analyzers for an analyzer that flags cases where an out-of-scope cancellation token is used in a method that has a cancellation token parameter.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:141]: Removed the special case handling here.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 15 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:98]: Removed the double-checked lock

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: This section is now removed.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 17]
```diff
@@ -93,41 +93,77 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
...
+                return null;
             }
+        }
 
-            var previousPackageSources = ImmutableInterlocked.InterlockedCompareExchange(ref _packageSources, packageSources, default);
-            if (previousPackageSources != null)
+        private async Task<ImmutableArray<PackageSource>> GetPackageSourcesImplAsync()
+        {
+            // Automatically retry the operation if it gets cancelled due to a change in the workspace.
+            while (true)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:131]: why do we need this?  if the workspace changed, wouldn't we simply compute the package sources the next time we get called?

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 18]
```diff
@@ -93,41 +93,77 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
...
             {
-                // Another thread already initialized _packageSources
-                packageSources = previousPackageSources;
+                var cancellationToken = _tokenSource.Token;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:133]: this does'nt feel safe to me.  _tokenSource is mutable data.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 19]
```diff
@@ -93,41 +93,77 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
...
+                try
+                {
+                    return await GetPackageSourcesImplAsync(cancellationToken).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    // The failure may have been caused by a workspace change; try again after a short delay
+                    await Task.Delay(TimeSpan.FromMilliseconds(250)).ConfigureAwait(false);
+                }
             }
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:143]: i don't think we need cancellation/retry here at all.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 17 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:131]: We don't have anything to return if the operation fails for this cancellation.

* Empty isn't valid, because it looks like a successful result even though it didn't complete.
* Cancellation exception isn't valid because the caller did not request it.

This retry loop is the easiest way to ensure the operation completes in a reasonable state.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 18 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:133]: `_tokenSource` is always valid, and `_tokenSource.Token` is immutable.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 19 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:143]: ➡️ This discussion is covered by https://github.com/dotnet/roslyn/pull/43718#discussion_r424001591

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 17 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs:131]: > Empty isn't valid, because it looks like a successful result even though it didn't complete.

Empty would be totally valid for me.  Note: we only cancel when a config change happens.  And in that case, we drop this task anyways.  So new callers will come in and see hte new task with the new values.  Any existing callers will see an empty array, which is 100% fine.

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs (hunk 18 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: This is now only read once for the life of the method.
